modules/lang-painless/src/main/java/org/elasticsearch/painless/DefMath.java

```diff
@@ -526,59 +526,63 @@ private static boolean eq(boolean a, boolean b) {
     }
 
     private static boolean eq(Object left, Object right) {
-        if (left != null && right != null) {
-            if (left instanceof Double) {
-                if (right instanceof Number) {
-                    return (double) left == ((Number) right).doubleValue();
-                } else if (right instanceof Character) {
-                    return (double) left == (char) right;
+        if (left == right) {
+            return true;
+        } else if (left != null && right != null) {
+            if (left.getClass() == right.getClass()) {
+                return left.equals(right);
```

Comment:
These lines are the new ones to short-circuit when they're the same instance or same type